kube/api-service.yaml

```
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+#
+# Copyright (C) 2023 KernelCI Authors.
+# Author: Sanskar Bhushan <sbdtu5498@gmail.com>
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: kernelci-api
+spec:
+  selector:
+    app: kernelci-api
+  ports:
+    - protocol: TCP
+      port: 8000
```

We are exposing `api` service on port 8001.
```suggestion
      port: 8001
```

---

Oh, service port and Local port are not the same thing. So service port is just being used to accumulate the traffic from all the target pods' ports and provide stable endpoints to them because pods are too fragile. Then this traffic can be forwarded to the local host via Kubectl port-forward. I hope this explanation makes it a bit clear. Let me know anything else you would like to discuss.

---

Yes that's possible. It can be done via ingresses as well or this is a type of service called ClusterIP that is used for internal networking. But if we have cloud provider based load balancers, we can also use LoadBalancer based service. It can be done in multiple different way. We can also forward the traffic to local instance (i.e. outside kubernetes cluster) via kubectl port-forward <svc-name> desired-port:8000 something like this.

---

Right, I think initially we would need to have something that's the most generic possible and not tailored to any particular Cloud provider or the definitive setup we'll have for kernelci.org.  So I would suggest to stick to this current approach as far as this PR is concerned with nothing in front of the port directly exposed by Kubernetes, which is only fine for a local setup but at least we'll get all this part tested and sorted out first.  I guess this should provide some reference for any Kubernetes-based deployment.

Then we can add a variant to set things up with SSL in Azure and also use the Atlas account for MongoDB rather than a local `db` service, but keep plain MongoDB in the reference config to not require Atlas as an external dependency - only as an option for those who which to use it.  This could then become the production-like deployment we would be using for beta-testing or "early access".

How does that sound?

---

Yeah, that's what I have thought. This deployment is already production ready. A few changes needs to be done. Not in API one in pipeline part there were a few security concerns such as DinD-based implementation, I have two more alternative. And for this one I have use mongo-data PVC as host path that is only for single node testing. When we will use Atlas we would have to change it a bit and use storage class that corresponds to Atlas. And then networking needs to be done, it would not require any heavylifting though. A bit of changes such as whether we would be using proxy such as nginx or envoy or can we use loadbalancers provided by Cloud providers themselves. But local testing is almost reaching it's final stage. Then we can move on to the next step soon.

---

> storage class that corresponds to Atlas

What do you mean by this?  If we use Atlas, we just don't run the `db` service at all but connect directly to Atlas.

---

Yeah, something like that. Basically, every storage providers provide ways to generically consume storage via Kubernetes API. So the volume controllers in Kubernetes take care of underlying intricacies. We just have to create persistent volume claims and when we will write persistent volume manifest that has storage details after we buy it and configure it on the cloud provider then we can provide those details in persistent volme manifest and yes, we are good to go. These two concepts persistent volume claims and persistent volume exist in Kubernetes to separate area of concerns of dev people and ops people. Administrator can buy storages handle secrets and provide them to a Kubernetes cluster via persistent volume and Developers can consume those storages using persistent volume claim. Hope this makes sense.

---

> Ah, I don't think we want to run Docker-in-Docker in Kubernetes. That's just a way to test the ability of the system to run jobs directly in Docker. So we can just remove the runner-docker service in the Kubernetes setup. The "native" runner also doesn't really make much sense in Kubernetes, we only need the `runner-k8s` and `runner-lava` ones in practice. Sorry if this wasn't clearly explained in the issue description.

Sure I will remove them.

---

> Oh, service port and Local port are not the same thing. So service port is just being used to accumulate the traffic from all the target pods' ports and provide stable endpoints to them because pods are too fragile. Then this traffic can be forwarded to the local host via Kubectl port-forward. I hope this explanation makes it a bit clear. Let me know anything else you would like to discuss.

Thanks for the clarification. Do you think Ingress or Gateway API can be useful here?

---

Oh sorry, I just saw your comment. Gateway API sure is very useful, but for basic traffic management, ingress would be sufficient enough. And yes we would be setting up an nginx ingress. Also, Gateway API is the unification of all the networking constructs that is possible in Infra management so if our use case is something like using service meshes and a bit more complicated, then we can also use Gateway API or service mesh. GatewayAPI has probably matured enough to be used in production so I guess going with any of the following i.e. service mesh or gateway API wouldn't be a bad choice.